### PR TITLE
[SOAR-4029] - Fix advanced query

### PR DESCRIPTION
**Pull request by jmcadams-r7**
## Proposed Changes

This PR splits the Advanced Query action in the InsightIDR plugin into two actions:
* Advanced Query on Log
* Advanced Query on Log Set

## Validation: 
```
RMT-MBP-5357:rapid7_insightidr jmcadams$ icon-validate .
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator ExampleInputValidator
[*] Plugin successfully validated!

----
[*] Total time elapsed: 813.831ms
```

## Testing: 
```
RMT-MBP-5357:rapid7_insightidr jmcadams$ ./run.sh -R tests/advanced_query_on_log.json -j
Running: cat tests/advanced_query_on_log.json | docker run --rm   -i rapid7/rapid7_insightidr:2.0.0  run | grep -- ^\{ | jq -r '.body | try(.log | split("\n") | .[]),.output'
Connect: Connecting...
rapid7/Rapid7 InsightIDR:2.0.0. Step name: advanced_query_on_log
Getting log entries from: https://us.api.insight.rapid7.com/log_search/management/logs
Starting new HTTPS connection (1): us.api.insight.rapid7.com:443
https://us.api.insight.rapid7.com:443 "GET /log_search/management/logs HTTP/1.1" 200 1120
Checking Active Directory against Endpoint Agents
Checking Active Directory against Alert Audit Log
Checking Active Directory against Log Updates
Checking Active Directory against Windows Defender
Checking Active Directory against Active Directory
Log found.
Found log with name Active Directory and ID: 5c6cc7cb-25bb-4423-8e43-7b2543489bbe
Getting logs from: https://us.api.insight.rapid7.com/log_search/query/logs/5c6cc7cb-25bb-4423-8e43-7b2543489bbe
Using parameters: {'query': 'where(agent,loose)', 'from': 1130778669000, 'to': 1610566806000}
https://us.api.insight.rapid7.com:443 "GET /log_search/query/logs/5c6cc7cb-25bb-4423-8e43-7b2543489bbe?query=where%28agent%2Cloose%29&from=1130778669000&to=1610566806000 HTTP/1.1" 202 309
Got a callback url. Polling results...
Trying to get results from callback URL: https://us.api.insight.rapid7.com/log_search/query/73099687-93d2-4fcf-a20f-e63b7415a6be:0:d62cc5a76678d24eb7a04713fb2d6a3bcd3ccb02:50:46625589017a70fbbf2ec3a2b9dc65cdee8cb8ee:
https://us.api.insight.rapid7.com:443 "GET /log_search/query/73099687-93d2-4fcf-a20f-e63b7415a6be:0:d62cc5a76678d24eb7a04713fb2d6a3bcd3ccb02:50:46625589017a70fbbf2ec3a2b9dc65cdee8cb8ee: HTTP/1.1" 200 263
Results were not ready. Sleeping 1 second and trying again.
Time left: 59
Progress: 0
https://us.api.insight.rapid7.com:443 "GET /log_search/query/73099687-93d2-4fcf-a20f-e63b7415a6be:0:d62cc5a76678d24eb7a04713fb2d6a3bcd3ccb02:50:46625589017a70fbbf2ec3a2b9dc65cdee8cb8ee: HTTP/1.1" 200 None
Sending results to orchestrator.
Connect: Connecting...
rapid7/Rapid7 InsightIDR:2.0.0. Step name: advanced_query_on_log
Getting log entries from: https://us.api.insight.rapid7.com/log_search/management/logs
Checking Active Directory against Endpoint Agents
Checking Active Directory against Alert Audit Log
Checking Active Directory against Log Updates
Checking Active Directory against Windows Defender
Checking Active Directory against Active Directory
Log found.
Found log with name Active Directory and ID: 5c6cc7cb-25bb-4423-8e43-7b2543489bbe
Getting logs from: https://us.api.insight.rapid7.com/log_search/query/logs/5c6cc7cb-25bb-4423-8e43-7b2543489bbe
Using parameters: {'query': 'where(agent,loose)', 'from': 1130778669000, 'to': 1610566806000}
Got a callback url. Polling results...
Trying to get results from callback URL: https://us.api.insight.rapid7.com/log_search/query/73099687-93d2-4fcf-a20f-e63b7415a6be:0:d62cc5a76678d24eb7a04713fb2d6a3bcd3ccb02:50:46625589017a70fbbf2ec3a2b9dc65cdee8cb8ee:
Results were not ready. Sleeping 1 second and trying again.
Time left: 59
Progress: 0
Sending results to orchestrator.

{
  "results": [
    {
      "labels": [],
      "timestamp": 1602662122928,
      "sequence_number": 3237133912949461000,
      "log_id": "5c6cc7cb-25bb-4423-8e43-7b2543489bbe",
      "message": {
        "timestamp": "2020-10-14T07:54:32.787Z",
        "destination_asset": "unknown",
        "source_asset_address": "10.4.22.248",
        "destination_asset_address": "10.4.22.248",
        "destination_account": "mhofert",
        "destination_domain": "soarcerrors",
        "destination_account_sid": "S-1-5-21-2677061338-3577197041-828036409-1605",
        "logon_type": "NETWORK",
        "result": "SUCCESS",
        "service": "advapi",
        "source_json": {
          "eventCode": 4624,
          "computerName": "iagent5-dcwin12.soarcerrors.vuln.lax.rapid7.com",
          "insertionStrings": [
            "S-1-5-18",
            "IAGENT5-DCWIN12$",
            "SOARCERRORS",
            "0x3e7",
            "S-1-5-21-2677061338-3577197041-828036409-1605",
            "mhofert",
            "SOARCERRORS",
            "0xac9b23c3",
            "3",
            "Advapi  ",
            "MICROSOFT_AUTHENTICATION_PACKAGE_V1_0",
            "IAGENT5-DCWIN12",
            "{00000000-0000-0000-0000-000000000000}",
            "-",
            "-",
            "0",
            "0x1f0",
            "C:\\Windows\\System32\\lsass.exe",
            "10.4.22.248",
            "50954",
            "%%1833"
          ],
          "timeGenerated": "20201014075432.787218-000"
        }
      },
      "links": [
        {
          "rel": "Context",
          "href": "https://us.api.insight.rapid7.com/log_search/query/context/3237133912949460992?per_page=50&timestamp=1602662122928&log_keys=5c6cc7cb-25bb-4423-8e43-7b2543489bbe&context_type=SURROUND"
        }
      ],
      "sequence_number_str": "3237133912949460992"
    },
    {
      "labels": [],
      "timestamp": 1602674915736,
      "sequence_number": 3237134602442961000,
      "log_id": "5c6cc7cb-25bb-4423-8e43-7b2543489bbe",
      "message": {
        "timestamp": "2020-10-14T11:28:25.865Z",
        "destination_asset": "unknown",
        "source_asset_address": "10.4.95.165",
        "destination_asset_address": "10.4.95.165",
        "destination_account": "administrator",
        "destination_domain": "soarcerrors",
        "destination_account_sid": "S-1-5-21-2677061338-3577197041-828036409-500",
        "logon_type": "NETWORK",
        "result": "SUCCESS",
        "service": "ntlmssp",
        "source_json": {
          "eventCode": 4624,
          "computerName": "iagent5-dcwin12.soarcerrors.vuln.lax.rapid7.com",
          "insertionStrings": [
            "S-1-0-0",
            "-",
            "-",
            "0x0",
            "S-1-5-21-2677061338-3577197041-828036409-500",
            "Administrator",
            "SOARCERRORS",
            "0xad2fe574",
            "3",
            "NtLmSsp ",
            "NTLM",
            "-",
            "{00000000-0000-0000-0000-000000000000}",
            "-",
            "NTLM V2",
            "128",
            "0x0",
            "-",
            "10.4.95.165",
            "38882",
            "%%1833"
          ],
          "timeGenerated": "20201014112825.865773-000"
        }
      },
      "links": [
        {
          "rel": "Context",
          "href": "https://us.api.insight.rapid7.com/log_search/query/context/3237134602442960896?per_page=50&timestamp=1602674915736&log_keys=5c6cc7cb-25bb-4423-8e43-7b2543489bbe&context_type=SURROUND"
        }
      ],
      "sequence_number_str": "3237134602442960896"
    },
    {
      "labels": [],
      "timestamp": 1602674915736,
      "sequence_number": 3237134602442965000,
      "log_id": "5c6cc7cb-25bb-4423-8e43-7b2543489bbe",
      "message": {
        "timestamp": "2020-10-14T11:28:25.716Z",
        "destination_asset": "unknown",
        "source_asset_address": "10.4.95.165",
        "destination_asset_address": "10.4.95.165",
        "destination_account": "administrator",
        "logon_type": "NETWORK",
        "result": "FAILED_BAD_PASSWORD",
        "service": "ntlmssp",
        "source_json": {
          "eventCode": 4625,
          "computerName": "iagent5-dcwin12.soarcerrors.vuln.lax.rapid7.com",
          "insertionStrings": [
            "S-1-0-0",
            "-",
            "-",
            "0x0",
            "S-1-0-0",
            "Administrator",
            "",
            "0xc000006d",
            "%%2313",
            "0xc000006a",
            "3",
            "NtLmSsp ",
            "NTLM",
            "-",
            "-",
            "-",
            "0",
            "0x0",
            "-",
            "10.4.95.165",
            "38860"
          ],
          "timeGenerated": "20201014112825.716896-000"
        }
      },
      "links": [
        {
          "rel": "Context",
          "href": "https://us.api.insight.rapid7.com/log_search/query/context/3237134602442964992?per_page=50&timestamp=1602674915736&log_keys=5c6cc7cb-25bb-4423-8e43-7b2543489bbe&context_type=SURROUND"
        }
      ],
      "sequence_number_str": "3237134602442964992"
    },
    ...<ommited for size>
```

```
RMT-MBP-5357:rapid7_insightidr jmcadams$ ./run.sh -R tests/advanced_query_on_log_set.json -j
Running: cat tests/advanced_query_on_log_set.json | docker run --rm   -i rapid7/rapid7_insightidr:2.0.0  run | grep -- ^\{ | jq -r '.body | try(.log | split("\n") | .[]),.output'
Connect: Connecting...
rapid7/Rapid7 InsightIDR:2.0.0. Step name: advanced_query_on_log_set
Getting log entries from: https://us.api.insight.rapid7.com/log_search/management/logsets
Starting new HTTPS connection (1): us.api.insight.rapid7.com:443
https://us.api.insight.rapid7.com:443 "GET /log_search/management/logsets HTTP/1.1" 200 819
Checking Asset Authentication against Host To IP Observations
Checking Asset Authentication against Internal Logs
Checking Asset Authentication against Unparsed Data
Checking Asset Authentication against Asset Authentication
Log set found.
Found log set with name Asset Authentication and ID: d89b6c00-ad44-48fc-a34c-cf0bee7bfbae
Getting logs from: https://us.api.insight.rapid7.com/log_search/query/logsets/d89b6c00-ad44-48fc-a34c-cf0bee7bfbae
Using parameters: {'query': 'where(result="SUCCESS", loose)', 'from': 1130778669000, 'to': 1610566854000}
https://us.api.insight.rapid7.com:443 "GET /log_search/query/logsets/d89b6c00-ad44-48fc-a34c-cf0bee7bfbae?query=where%28result%3D%22SUCCESS%22%2C+loose%29&from=1130778669000&to=1610566854000 HTTP/1.1" 202 351
Got a callback url. Polling results...
Trying to get results from callback URL: https://us.api.insight.rapid7.com/log_search/query/9f5163e7-9ae6-44be-836e-ddbb3129ead2:0:fe6e4e724b9cbb55a64547325c0a818d4aec1089:50:3540cb37a6fd21bed9a8189e04ea92cf3b79e526:
https://us.api.insight.rapid7.com:443 "GET /log_search/query/9f5163e7-9ae6-44be-836e-ddbb3129ead2:0:fe6e4e724b9cbb55a64547325c0a818d4aec1089:50:3540cb37a6fd21bed9a8189e04ea92cf3b79e526: HTTP/1.1" 200 302
Results were not ready. Sleeping 1 second and trying again.
Time left: 59
Progress: 0
https://us.api.insight.rapid7.com:443 "GET /log_search/query/9f5163e7-9ae6-44be-836e-ddbb3129ead2:0:fe6e4e724b9cbb55a64547325c0a818d4aec1089:50:3540cb37a6fd21bed9a8189e04ea92cf3b79e526: HTTP/1.1" 200 None
Sending results to orchestrator.
Connect: Connecting...
rapid7/Rapid7 InsightIDR:2.0.0. Step name: advanced_query_on_log_set
Getting log entries from: https://us.api.insight.rapid7.com/log_search/management/logsets
Checking Asset Authentication against Host To IP Observations
Checking Asset Authentication against Internal Logs
Checking Asset Authentication against Unparsed Data
Checking Asset Authentication against Asset Authentication
Log set found.
Found log set with name Asset Authentication and ID: d89b6c00-ad44-48fc-a34c-cf0bee7bfbae
Getting logs from: https://us.api.insight.rapid7.com/log_search/query/logsets/d89b6c00-ad44-48fc-a34c-cf0bee7bfbae
Using parameters: {'query': 'where(result="SUCCESS", loose)', 'from': 1130778669000, 'to': 1610566854000}
Got a callback url. Polling results...
Trying to get results from callback URL: https://us.api.insight.rapid7.com/log_search/query/9f5163e7-9ae6-44be-836e-ddbb3129ead2:0:fe6e4e724b9cbb55a64547325c0a818d4aec1089:50:3540cb37a6fd21bed9a8189e04ea92cf3b79e526:
Results were not ready. Sleeping 1 second and trying again.
Time left: 59
Progress: 0
Sending results to orchestrator.

{
  "results": [
    {
      "timestamp": 1602638719118,
      "labels": [],
      "sequence_number": 3237162474039750700,
      "log_id": "64e0f0e9-1a88-4601-ae36-f6d03687f313",
      "message": {
        "timestamp": "2020-10-14T01:24:48.212Z",
        "destination_asset": "iagent1-win10",
        "source_asset_address": "10.4.84.224",
        "destination_asset_address": "iagent1-win10",
        "destination_local_account": "administrator",
        "logon_type": "NETWORK",
        "result": "SUCCESS",
        "new_authentication": "false",
        "service": "ntlmssp ",
        "source_json": {
          "sourceName": "Microsoft-Windows-Security-Auditing",
          "insertionStrings": [
            "S-1-0-0",
            "-",
            "-",
            "0x0",
            "S-1-5-21-2947824804-3171091966-890232435-500",
            "Administrator",
            "IAGENT1-WIN10",
            "0x2bdd8edb",
            "3",
            "NtLmSsp ",
            "NTLM",
            "JCIFS1_1_C0",
            "{00000000-0000-0000-0000-000000000000}",
            "-",
            "NTLM V2",
            "0",
            "0x0",
            "-",
            "10.4.84.224",
            "53848",
            "%%1833",
            "-",
            "-",
            "-",
            "%%1843",
            "0x0",
            "%%1842"
          ],
          "eventCode": 4624,
          "computerName": "iagent1-win10",
          "sid": "",
          "isDomainController": false,
          "eventData": null,
          "timeWritten": "2020-10-14T01:24:47.206795000Z"
        }
      },
      "links": [
        {
          "rel": "Context",
          "href": "https://us.api.insight.rapid7.com/log_search/query/context/3237162474039750656?per_page=50&timestamp=1602638719118&log_keys=64e0f0e9-1a88-4601-ae36-f6d03687f313%3A5c6cc7cb-25bb-4423-8e43-7b2543489bbe&context_type=SURROUND"
        }
      ],
      "sequence_number_str": "3237162474039750656"
    },
    {
      "timestamp": 1602648672010,
      "labels": [],
      "sequence_number": 3237161573998805000,
      "log_id": "64e0f0e9-1a88-4601-ae36-f6d03687f313",
      "message": {
        "timestamp": "2020-10-14T04:10:22.482Z",
        "destination_asset": "ivm-console-test",
        "source_asset_address": "10.4.84.224",
        "destination_asset_address": "ivm-console-test",
        "destination_user": "user",
        "destination_account": "user",
        "destination_account_sid": "S-1-5-21-1663271794-3210079414-3662025226-1000",
        "logon_type": "NETWORK",
        "result": "SUCCESS",
        "new_authentication": "false",
        "service": "ntlmssp ",
        "source_json": {
          "sourceName": "Microsoft-Windows-Security-Auditing",
          "insertionStrings": [
            "S-1-0-0",
            "-",
            "-",
            "0x0",
            "S-1-5-21-1663271794-3210079414-3662025226-1000",
            "user",
            "IVM-CONSOLE-TES",
            "0x705e7cbb",
            "3",
            "NtLmSsp ",
            "NTLM",
            "",
            "{00000000-0000-0000-0000-000000000000}",
            "-",
            "NTLM V2",
            "128",
            "0x0",
            "-",
            "10.4.84.224",
            "39164"
          ],
          "eventCode": 4624,
          "computerName": "ivm-console-test",
          "sid": "",
          "isDomainController": false,
          "eventData": null,
          "timeWritten": "2020-10-14T04:10:22.088438200Z"
        }
      },
      "links": [
        {
          "rel": "Context",
          "href": "https://us.api.insight.rapid7.com/log_search/query/context/3237161573998804992?per_page=50&timestamp=1602648672010&log_keys=64e0f0e9-1a88-4601-ae36-f6d03687f313%3A5c6cc7cb-25bb-4423-8e43-7b2543489bbe&context_type=SURROUND"
        }
      ],
      "sequence_number_str": "3237161573998804992"
    },
    ... <ommitted for size>
```

### CODE REVIEW

File: rapid7_insightidr/plugin.spec.yaml [hunk 1]
```diff
@@ -599,22 +599,61 @@ actions:
...
+        title: Time From
+        description: Beginning time and date for the query
+        type: string
+        example: "01-01-2020"
+        required: true
+      time_to:
+        title: Time To
+        description: Ending date and time for the query. If this is left blank, the current time will be used
+        type: string
+        example: "12-31-2020"
```
Comment (jschipp-r7) [inline rapid7_insightidr/plugin.spec.yaml:629]: This example doesn't have a time but states a date and time is supported

File: rapid7_insightidr/plugin.spec.yaml [hunk 2]
```diff
@@ -599,22 +599,61 @@ actions:
...
+        title: Query
+        description: Query
+        type: string
+        example: "where(user=adagentadmin, loose)"
+        required: true
+      time_from:
+        title: Time From
+        description: Beginning time and date for the query
+        type: string
+        example: "01-01-2020"
```
Comment (jschipp-r7) [inline rapid7_insightidr/plugin.spec.yaml:623]: This example doesn't have a time but states a date and time is supported. As a user I also can't tell how to specify a time and if a time is required.

File: rapid7_insightidr/help.md [hunk 3]
```diff
@@ -36,15 +36,53 @@ Example input:
...
+
+##### Output
+
+|Name|Type|Required|Description|
+|----|----|--------|-----------|
+|results|[]events|True|Query Results|
+
+Example output:
+
+```
```
Comment (jschipp-r7) [inline rapid7_insightidr/help.md:73]: Missing output example

File: rapid7_insightidr/plugin.spec.yaml [hunk 4]
```diff
@@ -599,22 +599,61 @@ actions:
...
+        type: "[]events"
+        required: true
+
+  advanced_query_on_log_set:
+    title: Advanced Query on Log Set
+    description: Realtime query into an Insight IDR log set
+    input:
+      query:
+        title: Query
+        description: Query
```
Comment (jschipp-r7) [inline rapid7_insightidr/plugin.spec.yaml:615]:
```suggestion
        description: LQL Query
```

File: rapid7_insightidr/help.md [hunk 5]
```diff
@@ -36,15 +36,111 @@ Example input:
 
 ### Actions
 
-#### Advanced Query
+#### Advanced Query on Log Set
 
-This action is used to run realtime queries into Insight IDR logs.
+This action is used to realtime query into an Insight IDR log set.
```
Comment (jschipp-r7) [inline rapid7_insightidr/help.md:41]: Would be good to explain when to use Log vs Logset version of the action in each action's description. The names are very similar and a new IDR user may not know

**Review by jschipp-r7 — CHANGES_REQUESTED** [review]

File: rapid7_insightidr/plugin.spec.yaml [hunk 6]
```diff
@@ -564,27 +564,27 @@ actions:
...
-        example: "01-01-2020"
+        example: "01-01-2020T00:00:00"
         required: true
       time_to:
         title: Time To
-        description: Ending date and time for the query. If this is left blank, the current time will be used
+        description: Ending date and time for the query. If this is left blank, the current time will be used. The format is flexible and will work with simple dates to full ISO time
         type: string
-        example: "12-31-2020"
+        example: "12-31-202012-31-2020T00:00:00"
```
Comment (jschipp-r7) [inline rapid7_insightidr/plugin.spec.yaml:587]: Date is duplicated in this one

File: rapid7_insightidr/plugin.spec.yaml [hunk 7]
```diff
@@ -564,27 +564,27 @@ actions:
...
         title: Query
-        description: Query
+        description: LQL Query
         type: string
         example: "where(user=adagentadmin, loose)"
         required: true
       time_from:
         title: Time From
-        description: Beginning time and date for the query
+        description: Beginning time and date for the query. The format is flexible and will work with simple dates to full ISO time
```
Comment (jschipp-r7) [inline rapid7_insightidr/plugin.spec.yaml:579]: We should be showing examples of both on all these descriptions

```suggestion
        description: Beginning time and date for the query. The format is flexible and will work with simple dates (e.g. 01-01-2020) to full ISO time (e.g. 01-01-2020T00:00:00)
```

File: rapid7_insightidr/help.md [hunk 8]
```diff
@@ -81,7 +81,103 @@ Example output:
...
+      ],
+      "sequence_number_str": "123456789123456789"
+    }
+  ]
+}
+```
+
+#### Advanced Query on Log
+
+This action is used to realtime query into Insight IDR logs.
```
Comment (jschipp-r7) [inline rapid7_insightidr/help.md:137]: Document when / why to use compare to Log Set

File: rapid7_insightidr/help.md [hunk 9]
```diff
@@ -36,18 +36,18 @@ Example input:
 
 ### Actions
 
-#### Advanced Query
+#### Advanced Query on Log Set
 
-This action is used to run realtime queries into Insight IDR logs.
+This action is used to realtime query into an Insight IDR log set.
```
Comment (jschipp-r7) [inline rapid7_insightidr/help.md:41]: Document when / why to use compare to Log action

File: rapid7_insightidr/help.md [hunk 10]
```diff
@@ -552,6 +648,7 @@ _This plugin does not contain any troubleshooting information._
 
 # Version History
 
+* 2.0.0 - Update to refactor advanced query. Split Advanced Query into two new actions Advanced Query on Log and Advanced Query on Log Set
```
Comment (jschipp-r7) [inline rapid7_insightidr/help.md:651]:
```suggestion
* 2.0.0 - Refactor and split Advanced Query into two new actions Advanced Query on Log and Advanced Query on Log Set
```

**Review by jschipp-r7 — CHANGES_REQUESTED** [review]

File: rapid7_insightidr/help.md (hunk 10 above)
Comment (jmcadams-r7) [reply rapid7_insightidr/help.md:651]: Update to is part of the style guide BTW.

Comment (jschipp-r7) [reply rapid7_insightidr/help.md]: Yeah, wanted to consolidate it into a single sentence for readability

**Review by jschipp-r7 — APPROVED** [review]

**Review by ncramer-r7 — APPROVED** [review]

**Pull request merged**